Sync/BatchingClient.swift

```diff
@@ -0,0 +1,218 @@
+    private var totalRecords: Int = 0
+
+    // Keep track of the limits of a single POST
+    private var postBytes: ByteCount = 0
+    private var postRecords: Int = 0
+
+    private var records = [UploadRecord]()
+
+    private var onCollectionUploaded: (POSTResult, Timestamp?) -> DeferredTimestamp
+
+    private func batchQueryParamWithValue(value: String) -> NSURLQueryItem {
+        return NSURLQueryItem(name: "batch", value: value)
+    }
+
+    init(config: InfoConfiguration, ifUnmodifiedSince: Timestamp? = nil, serializeRecord: (Record<T>) -> String?,
+         uploader: BatchUploadFunction, onCollectionUploaded: (POSTResult, Timestamp?) -> DeferredTimestamp) {
+        self.config = config
+        self.ifUnmodifiedSince = ifUnmodifiedSince
+        self.uploader = uploader
+        self.serializeRecord = serializeRecord
+
+        self.onCollectionUploaded = onCollectionUploaded
+    }
+
+    public func endBatch() -> Success {
+        guard !records.isEmpty else {
+            return succeed()
+        }
+
+        if let token = self.batchToken {
+            return commitBatch(token) >>> succeed
+        } else {
+            let lines = self.freezePost()
+            let queryParams = [
+                batchQueryParamWithValue("true"),
+                NSURLQueryItem(name: "commit", value: "true")
+            ]
+            return self.uploader(lines: lines, ifUnmodifiedSince: self.ifUnmodifiedSince, queryParams: queryParams)
+                >>== effect(moveForward)
+                >>> succeed
+        }
+    }
+
+    public func addRecords(records: [Record<T>]) -> Success {
+        guard !records.isEmpty else {
+            return succeed()
+        }
+
+        do {
+            let serialized = try records.map { try serialize($0) }
+            return addRecords(serialized.generate())
+        } catch let e {
+            return deferMaybe(e as! MaybeErrorType)
+        }
+    }
+
+    private func addRecords(generator: IndexingGenerator<[UploadRecord]>) -> Success {
+        var mutGenerator = generator
+        while let record = mutGenerator.next() {
+            if let uploadOp = accumulateRecord(record) {
+                return uploadOp >>> {
+                    if let deferred = self.accumulateRecord(record) {
+                        return deferred >>> { self.addRecords(mutGenerator) }
+                    }
+                    return self.addRecords(mutGenerator)
+                }
+            }
+        }
+        return succeed()
+    }
+
+    private func accumulateRecord(record: UploadRecord) -> DeferredResponse? {
+        if let token = self.batchToken {
+            if fitsInBatch(record) {
+                if addToPost(record) {
+                    // Only count the record towards the batch if we could add it to a payload...
+                    addToBatch(record)
+                    return nil
+                } else {
+                    // Otherwise, we don't added the record and just POST away.
```

Comment:
Oh, is this the source of that confusing double-attempt above? If so, please tidy that up (perhaps moving it here) or comment `accumulateRecord` to explain!